BUG: Fix for #10533 np.dtype(ctype) does not respect endianness

bmakos opened the pull request:
See: #10533
Added some code that uses the __name__ of the ctypes type. Also
added some new test to make sure we'll know if the ctypes data
that allows this workaround breaks in the future.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

bmakos committed:
BUG: Fix for #10533 np.dtype(ctype) does not respect endianness

See: #10533
Added some code that uses the __name__ of the ctypes type. Also
added some new test to make sure we'll know if the ctypes data
that allows this workaround breaks in the future.

bmakos committed:
Fix for #10533 np.dtype(ctype) does not respect endianness

Removed the TODO for this bug.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -61,12 +60,24 @@ def _from_ctypes_structure(t):
         fields = []
         for fname, ftyp in t._fields_:
-            fields.append((fname, dtype_from_ctypes_type(ftyp)))
+            fields.append((fname, dtype_from_ctypes_endianness(ftyp)))
```
There's no need to change this line or the one above - the recursion will handle it just fine

bmakos replied:
Indeed. Changed.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -61,12 +60,24 @@ def _from_ctypes_structure(t):
+    """
+    Return the dtype type with endianness included if it's the case
+    """
+    if t.__name__.endswith('_be'):
```
Found something better - `if t.__ctypes_be__ is t`

bmakos replied:
I don't follow. 
The ctype type has both `_be` and `_le` in it's `__dict__ `independent of actual endianness.

bmakos committed:
Fix for #10533 np.dtype(ctype) does not respect endianness

Solved the recursiveness issue that failed the automated test.
Also did some of Eric's requested changes.

bmakos commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -61,12 +60,24 @@ def _from_ctypes_structure(t):
         return np.dtype(fields, align=True)
 
 
+def dtype_from_ctypes_endianness(t):
```
Ok, changed.

bmakos committed:
Small mistake in the new function fixed.

bmakos commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -82,6 +93,8 @@ def dtype_from_ctypes_type(t):
         raise NotImplementedError(
             "conversion from ctypes.Union types like {} to dtype"
             .format(t.__name__))
+    elif isinstance(t, type(ctypes.c_int)): # Could be any simple type instead of c_int, all return the same type
```
Done.

bmakos replied:
Done now, missed it in previous commit.

bmakos committed:
Changed 'if t.__name__.endswith('_be') ' with
'if t.__ctypes_be__ is t' on Eric's suggestion.

bmakos commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -67,6 +66,18 @@ def _from_ctypes_structure(t):
+    elif t.__ctype_le__ is t:
+        return '<' + t._type_
+    else:
+        return t._type_
```
Yes, did this initially but then changed it for some reason. Anyhow, it indeed works and it's consistent with other def's.

bmakos commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -82,6 +93,8 @@ def dtype_from_ctypes_type(t):
+    elif isinstance(t, type(ctypes.c_int)): # Could be any simple type instead of c_int, all return the same type
+        return np.dtype(dtype_from_ctypes_scalar(t))
     elif isinstance(t._type_, str):
         return np.dtype(t._type_)
```
Done I think.

bmakos commented on numpy/core/tests/test_dtype.py:
```diff
@@ -856,8 +856,19 @@ class PackedStructure(ctypes.Structure):
-        reason="non-native endianness does not work - see gh-10533")
+    def test_complex_big_endian_structure_packed(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
+            _fields_ = [('one', ctypes.c_uint8.__ctype_be__),('two', ctypes.c_uint32.__ctype_be__)]
```
Fixed.

bmakos commented on numpy/core/tests/test_dtype.py:
```diff
@@ -883,3 +893,29 @@ class PaddedStruct(ctypes.BigEndianStructure):
+    def test_big_endian_simple_type(self):
+        byte_order = np.dtype(ctypes.c_uint16.__ctype_be__).byteorder
+        if sys.byteorder == 'little':
+           expected = '>'
```
Fixed.

bmakos commented on numpy/core/tests/test_dtype.py:
```diff
@@ -856,8 +856,19 @@ class PackedStructure(ctypes.Structure):
+
+    def test_complex_big_endian_structure(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
+            _fields_ = [('one', ctypes.c_uint16.__ctype_be__),('two', ctypes.c_uint32.__ctype_be__)]
```
Removed.

bmakos commented on numpy/core/tests/test_dtype.py:
```diff
@@ -883,3 +893,29 @@ class PaddedStruct(ctypes.BigEndianStructure):
+           expected = '<'
+        else:
+            expected = '='
+        assert byte_order == expected
```
Merged into 1 test.

eric-wieser replied:
May as well remove the old test above this comment

bmakos replied:
Yes, missed to do it. Done now.

bmakos committed:
Code cleanup on Eric's recommendations.

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -883,3 +886,17 @@ class PaddedStruct(ctypes.BigEndianStructure):
+        assert_equal(np.dtype(ctypes.c_uint16.__ctype_le__), np.dtype('<u2'))
+        assert_equal(np.dtype(ctypes.c_uint16.__ctype_be__), np.dtype('>u2'))
+        assert_equal(np.dtype(ctypes.c_uint8.__ctype_le__), np.dtype('u1'))
+        assert_equal(np.dtype(ctypes.c_uint8.__ctype_be__), np.dtype('u1'))
```
You should actually use `self.check` here, it tests both types and instances.

bmakos replied:
Done.

bmakos replied:
Ok, I just had to look how `check `was defined... Did the commit now.

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -856,8 +856,13 @@ class PackedStructure(ctypes.Structure):
-        reason="non-native endianness does not work - see gh-10533")
+    def test_complex_big_endian_structure_packed(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
+            _fields_ = [('one', ctypes.c_uint8),('two', ctypes.c_uint32)]
```
Can you line wrap this like in the tests below?

bmakos replied:
Yes, done.

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -856,8 +856,13 @@ class PackedStructure(ctypes.Structure):
 
-    @pytest.mark.xfail(sys.byteorder != 'little',
-        reason="non-native endianness does not work - see gh-10533")
+    def test_complex_big_endian_structure_packed(self):
```
I'd drop the word "complex" - this is no more complex than the test below it.

bmakos replied:
Done.

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -856,8 +856,13 @@ class PackedStructure(ctypes.Structure):
-    @pytest.mark.xfail(sys.byteorder != 'little',
-        reason="non-native endianness does not work - see gh-10533")
+    def test_complex_big_endian_structure_packed(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
```
Can you add a test with a packed LittleEndianStructure too?

bmakos replied:
Added.

bmakos committed:
Cleaned up the tests on Eric's suggestions.

eric-wieser reviewed (approved):
Looks great, thanks! I plan to squash together all the commits when I merge

bmakos commented:
Thank you for the review Eric! I enjoyed working on this. I'll go choose some other bug today to check out.

eric-wieser commented:
If @euler16 doesn't pick it back up, I'm sure you'd have an easy time of #12273. On the one hand, it'd be great to leave them a chance to contribute - but on the other hand, I'd really like to be able to say that 1.16 has full ctypes support. Let's give them a few more days to reply...

charris commented:
Thanks @bmakos .
